### PR TITLE
feat: check the signature on updating public address of the node

### PR DESCRIPTION
**Pull request by Cifko**
Check the signature when calling update public address for a node.
- check if the signature is valid
- check if the signature is for the same small id that's in the request

### CODE REVIEW

File: atoma-proxy/src/server/http_server.rs [hunk 1]
```diff
@@ -263,8 +275,90 @@ pub(crate) struct NodePublicAddressRegistrationOpenApi;
...
+        .map_err(|err| {
+            error!("Failed to get node Sui address: {:?}", err);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?
+        .ok_or_else(|| {
+            error!("Node Sui address not found");
+            StatusCode::NOT_FOUND
+        })?;
+
+    // Check if the small ID in the request matches the Sui address in the signature.
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/http_server.rs:356]: This is not a small id check !

File: atoma-proxy/src/server/http_server.rs [hunk 2]
```diff
@@ -263,8 +275,90 @@ pub(crate) struct NodePublicAddressRegistrationOpenApi;
...
+    if node_sui_address != sui_address.to_string() {
+        error!("Node Sui address does not match the node s");
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/http_server.rs:358]: The error message seems to be incomplete

File: atoma-proxy/src/server/http_server.rs [hunk 3]
```diff
@@ -263,8 +275,90 @@ pub(crate) struct NodePublicAddressRegistrationOpenApi;
...
+        })?;
+    let sui_address = SuiAddress::from(&public_key);
+
+    let mut blake2b_hash = Blake2b::new();
+    blake2b_hash.update(&body_bytes);
+    let body_blake2b_hash: GenericArray<u8, U32> = blake2b_hash.finalize();
+    let body_blake2b_hash_bytes: [u8; 32] = body_blake2b_hash
+        .as_slice()
+        .try_into()
+        .expect("Invalid Blake2b hash length");
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/http_server.rs:319]: So we really want to panic here ?

File: atoma-state/src/state_manager.rs [hunk 4]
```diff
@@ -2160,23 +2162,17 @@ impl AtomaState {
...
-                    (node_small_id, public_address) 
-                    VALUES ($1, $2) 
-                    ON CONFLICT (node_small_id)
-                    DO UPDATE SET public_address = EXCLUDED.public_address",
-        )
-        .bind(small_id)
-        .bind(address)
-        .execute(&self.db)
-        .await?;
+        sqlx::query("UPDATE nodes SET public_address = $2 WHERE node_small_id = $1")
```
Comment (jorgeantonio21) [inline atoma-state/src/state_manager.rs:2165]: I do not understand how we can update a node address, isn't this its public key ?

**Review by jorgeantonio21 — CHANGES_REQUESTED** [review]
Looking good, but some changes are required

File: atoma-proxy/src/server/http_server.rs (hunk 2 above)
Comment (Cifko) [reply atoma-proxy/src/server/http_server.rs:358]: Changed

File: atoma-proxy/src/server/http_server.rs (hunk 3 above)
Comment (Cifko) [reply atoma-proxy/src/server/http_server.rs:319]: no

File: atoma-state/src/handlers.rs [hunk 5]
```diff
@@ -607,6 +607,18 @@ pub(crate) async fn handle_stack_attestation_dispute_event(
     Ok(())
 }
 
+pub(crate) async fn handle_node_registration_event(
```
Comment (jorgeantonio21) [inline atoma-state/src/handlers.rs:610]: add docs and instrument

File: atoma-state/src/handlers.rs [hunk 6]
```diff
@@ -607,6 +607,18 @@ pub(crate) async fn handle_stack_attestation_dispute_event(
...
+    state_manager: &AtomaStateManager,
+    event: NodeRegisteredEvent,
+    address: String,
+) -> Result<()> {
+    state_manager
+        .state
+        .insert_new_node(event.node_small_id.inner as i64, address)
+        .await?;
```
Comment (jorgeantonio21) [inline atoma-state/src/handlers.rs:618]: log potential error here

File: atoma-proxy/src/server/http_server.rs [hunk 7]
```diff
@@ -297,8 +309,91 @@ pub(crate) struct NodePublicAddressRegistrationOpenApi;
 #[instrument(level = "info", skip_all)]
 pub async fn node_public_address_registration(
     State(state): State<ProxyState>,
-    Json(payload): Json<NodePublicAddressAssignment>,
+    headers: HeaderMap,
+    request: Request<Body>,
 ) -> Result<Json<Value>, StatusCode> {
+    let base64_signature = headers
+        .get("X-Signature")
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/http_server.rs:316]: add the right constant

File: atoma-proxy/src/server/http_server.rs [hunk 8]
```diff
@@ -297,8 +309,91 @@ pub(crate) struct NodePublicAddressRegistrationOpenApi;
...
+    let public_key_bytes = signature.public_key_bytes();
+    let public_key =
+        SuiPublicKey::try_from_bytes(signature.scheme(), public_key_bytes).map_err(|e| {
+            error!("Failed to extract public key from bytes, with error: {e}");
+            StatusCode::BAD_REQUEST
+        })?;
+    let sui_address = SuiAddress::from(&public_key);
+
+    let mut blake2b_hash = Blake2b::new();
+    blake2b_hash.update(&body_bytes);
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/http_server.rs:348]: what is the body look like ? can you give an example ?

Also should the node repository have a method to call this endpoint ?

File: atoma-proxy/src/server/http_server.rs [hunk 9]
```diff
@@ -297,8 +309,91 @@ pub(crate) struct NodePublicAddressRegistrationOpenApi;
...
+    let body_blake2b_hash: GenericArray<u8, U32> = blake2b_hash.finalize();
+    let body_blake2b_hash_bytes: [u8; 32] =
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/http_server.rs:350]: replace `32`with proper constant

**Review by jorgeantonio21 — COMMENTED** [review]
Looking better but still some changes are required.

I do believe we need a p2p network so that we don't need to call this method, as there might be 1000s of proxies and node operators can't possible register to all of them...

**Comment by Cifko** [comment]
> Looking better but still some changes are required.
> 
> I do believe we need a p2p network so that we don't need to call this method, as there might be 1000s of proxies and node operators can't possible register to all of them...

I agree, but currently we can't deploy working network without this. And we don't have p2p implemented. So if we don't plant o hardcode the addresses into the proxy right now, we need this pr to go in.

File: atoma-state/src/handlers.rs (hunk 6 above)
Comment (Cifko) [reply atoma-state/src/handlers.rs:618]: the errors are propagated all the way to error `handle_tasks_results` in `main.rs` We should not double log errors.

File: atoma-proxy/src/server/http_server.rs (hunk 8 above)
Comment (Cifko) [reply atoma-proxy/src/server/http_server.rs:348]: yes we should
The body is of type `NodePublicAddressAssignment`
```
pub struct NodePublicAddressAssignment {
    /// Unique small integer identifier for the node
    node_small_id: u64,
    /// The public address of the node
    public_address: String,
}
```

**Review by jorgeantonio21 — APPROVED** [review]
LGTM

**Pull request merged**